src/reducers/operations.ts

```diff
@@ -0,0 +1,120 @@
+import { Action, Middleware } from "redux";
+import * as objectHash from 'object-hash';
+import * as update from "immutability-helper";
+import * as moment from 'moment';
+import { isDismissOperation } from 'actions/operationsActions';
+import { AsyncActionSequence, IAsyncAction , isAsyncAction } from 'actions/async';
+
+export enum OperationsStatus {
+  Pending = 'pending',
+  Failure = 'failure',
+  Success = 'success'
+}
+
+export interface IOperation {
+  status: OperationsStatus;
+  message: string;
+  step: number;
+  totalSteps?: number;
+  timestamp: number;
+}
+
+/**
+ * Each operation is given a stable key which is simply the hash of it's action's `type` and `meta`.
+ * This means that that actions with the same `type` and different `meta` will be treated as different.
+ */
+export interface IOperationsState {
+  [hash: string]: IOperation
+}
+
+/**
+ * A reducer that keeps track of currently running operations.
+ * Multiple `pending` actions with the same key will increment the `step` until either a `failure`, a `succeess` or a `reset` happens.
+ */
+export const operationsReducer =
+  (state: IOperationsState = {}, action: Action): IOperationsState => {
+      if (isAsyncAction(action)) {
+        const { type, meta, sequence, operation } = action;
+        const hash = objectHash({ type, meta });
+
+        const operationsConfig = operation || {};
+
+        const defaults: IOperation = {
+          status: OperationsStatus.Pending,
+          message: operationsConfig.message || `Operation #${hash.substr(0, 4)}`,
+          step: 0,
+          totalSteps: operationsConfig.totalSteps,
+          timestamp: +moment() // TODO: this makes the reducer impure. figure out a better way.
+        }
+
+        /* initialize with defaults if does not exist yet */
+        if (!state[hash]) {
+          state = update(state, {
+            [hash]: {
+              $set: defaults
+            }
+          });
+        }
+
+        switch (action.sequence) {
+          case 'reset':
+          case 'success':
+            return update(state, { $unset: [hash] });
+          case 'pending':
+            return update(state, {
+              [hash]: {
+                $merge: {
+                  status: OperationsStatus.Pending,
+                  step: state[hash].status === OperationsStatus.Pending ? state[hash].step + 1 : 0,
```

Comment:
im a little unclear how we will use this step functionality. steps for us are multiple blockchain transactions, its not like we will be dispatching multiple actions for each one

Reply:
Steps increment for each `pending` action. So if you have multiple transactions, you would dispatch a `pending` action after completing each transaction.

Reply:
and you are saying that we cant really do that right now. seems like we need to ask MetaMask for transaction results somehow

Reply:
yeah that would be nice, lets think through how to articulate it to Doug. notifications on completed transactions somehow

Reply:
Not sure what is a good solution for this. A callback? A global `pendingTransactions` function for all transactions? Is this going to be needed anywhere except token approval?

Maybe we should just handle the token approval in alchemy instead of Arc.js.

What do you think @dkent600 @tibetsprague ?